File: web/packages/lib/src/models/tick.ts
import * as z from "zod";
import type { Tick } from "./generated";

export const tickSchema: z.ZodType<Tick> = z.object({
  id: z.number(),
  timestamp: z.coerce.date(),
  problemId: z.number(),
  zone1: z.boolean(),
  attemptsZone1: z.number(),
  zone2: z.boolean(),
  attemptsZone2: z.number(),
  top: z.boolean(),
  attemptsTop: z.number(),
});

File: backend/internal/usecases/validators/tick.go
package validators

import (
	"github.com/climblive/platform/backend/internal/domain"
	"github.com/go-errors/errors"
)

var errTickConstraintViolation = errors.New("constraint violation")

type TickValidator struct {
}

func (v TickValidator) Validate(tick domain.Tick) error {
	switch {
	case tick.AttemptsZone1 > tick.AttemptsZone2:
		fallthrough
	case tick.AttemptsZone2 > tick.AttemptsTop:
		fallthrough
	case tick.Top && !tick.Zone1:
		fallthrough
	case tick.Top && !tick.Zone2:
		fallthrough
	case tick.Zone2 && !tick.Zone1:
		return errors.Errorf("%w: %w", domain.ErrInvalidData, errTickConstraintViolation)
	}

	return nil
}

func (v TickValidator) IsValidationError(err error) bool {
	return errors.Is(err, errTickConstraintViolation)
}
